0000-experiment-dataframe.md

```diff
@@ -0,0 +1,404 @@
+
+#### Example 1: Parallel experiment.
+
+```python
+pexp = ParallelExperiment(
+  [
+    T1(qubit=0, delays=delays),
+    T1(qubit=1, delays=delays),    
+  ],
+  backend=ibm_washington,
+)
+experment_data = pexp.run()
+
+experiment_data.analysis_results(dataframe=True)
+```
+
+Analysis results:
+
+|          | name   | value  | quality | components |  backend        | run_time                   | experiment | ... |
+|----------|--------|--------|---------|------------|-----------------|----------------------------|------------|-----|
+| 03fe91s1 | T1     | 100e-6 | good    | [Q0]       |  ibm_washington | 2022-11-22 03:55:30.313872 | T1         | ... |
+| 1c4ad2a1 | T1     | 120e-6 | good    | [Q1]       |  ibm_washington | 2022-11-22 03:55:30.313872 | T1         | ... |
+
+
+#### Example 2: Batch experiment.
+
+```python
+bexp = BatchExperiment(
+  [
+    T1(qubit=0, delays=delays),
+    T2Ramsey(qubit=0, delays=delays),    
+  ],
+  backend=ibm_washington,
+)
+experment_data = bexp.run()
+
+experiment_data.analysis_results(dataframe=True)
+```
+
+Analysis results:
+
+|          | name   | value  | quality | components |  backend        | run_time                   | experiment | ... |
+|----------|--------|--------|---------|------------|-----------------|----------------------------|------------|-----|
+| 17636b47 | T1     | 100e-6 | good    | [Q0]       |  ibm_washington | 2022-11-22 03:58:36.523230 | T1         | ... |
+| 2f50a741 | T2star |  80e-6 | good    | [Q0]       |  ibm_washington | 2022-11-22 03:58:36.523230 | T2Ramsey   | ... |
+
+
+#### Example 3: Repeating the same experiment.
+
+```python
+exp = T1(qubit=0, delays=delays)
+
+trend_data = []
+for _ in range(2):
+  experment_data = exp.run()
+  trend_data.append(experment_data)
+  time.sleep(3600)
+
+# combine trend data
+...
+```
+
+Analysis results (combined):
+
+|          | name   | value  | quality | components |  backend        | run_time                   | experiment | ... |
+|----------|--------|--------|---------|------------|-----------------|----------------------------|------------|-----|
+| 03fe91s1 | T1     | 100e-6 | good    | [Q0]       |  ibm_washington | 2022-11-22 04:01:30.313872 | T1         | ... |
+| 1c4ad2a1 | T1     | 103e-6 | good    | [Q0]       |  ibm_washington | 2022-11-22 05:01:14.123201 | T1         | ... |
+
+
+#### Complement
+
+As these examples show, experimentalist can distinguish the outcomes of batch (with `experiment`) and parallel experiments (with `components`). Even though there are multiple entries for the identical parameter (example 3), they have different result and experiment IDs and run times. In other words, one can sort the entries by `run_time` and get the time-ordered sequence of `value` for the trend analysis:
```

Comment:
I'm thinking about cases of complicated nested composite experiments where looking at `experiment` and `components` are not enough to distinguish between different results. The user might be interested in is something like "I want the result of the fourth T1 experiment that was run simultaneously on Q0 and Q10." Of course, this can be deduced by looking at all the columns, but are we expecting the users to be familiar with pandas and write their own multi-column filters?

Reply:
Good point. Such functionality can be provided through the pandas accessor, but the good point of using pandas is they can find many well written documentation in the web. As you know we don't have enough capability of maintaining detailed documentation and I'd like to avoid overengineering at this stage (even if we provide the logic it is tough for users to actually use it). Can you provide some configuration that might mess up the result table?

Reply:
Yeah, I don't think we should make more work for ourselves implementing too many custom accessors since pandas already has many features and is well-documented. Perhaps a tutorial showing some particularly useful operations would be good enough. As we discussed, the main issue I can think of is `run_time` can be potentially out of order from the actual execution time on the device. I can think of a few times that are relevant:
1. time when job was submitted to the queue
2. time when job actually started to run
3. time when job finished running
4. `run_time` as defined here, time of experiment data receiving the job results

Usually 1 and 2 will be in the same order, but experimentalists who have access to the queue can potentially move jobs around in it. So for something like "get the result of the fourth T1 experiment", I wonder which time metric we should be using, or should we be recording multiple times and ask the user to specify what they want.

Reply:
Maybe it is worth thinking a bit more about how we should record times.

For experiments I would expect the important times to be 2 and 3, when it started running on the device, and when it finished running on the device. However these times are generally specific to backend data (count/iq), not individual analysis results so any results processed from the same batch of data, such as parallel experiments, will have the same times. I don't think the specific time that ExperimentData received a job result is particular relevant (unless the job contains no timing information for its execution).

Another idea would be to include a `created_time` for analysis results like you have for artifacts. I'm imagining this could be relevant if you re-ran analysis on exisiting data (which would have a fixed time for when it ran) and wanted to save both results to the DB (im not sure this is allowed through qiskit-experiments without the two results also having different experiment ids, but you could probably do it directly through service).

Reply:
Thanks Chris. I noticed that Job object reports multiple timestep, i.e.
```python
job = backend.run(qc)
job._wait_for_completion()
job.time_per_step()
```
this returns a dictionary
```
{'CREATING': datetime.datetime(2023, 1, 26, 15, 22, 4, 465000, tzinfo=tzlocal()),
 'CREATED': datetime.datetime(2023, 1, 26, 15, 22, 5, 813000, tzinfo=tzlocal()),
 'VALIDATING': datetime.datetime(2023, 1, 26, 15, 22, 6, 43000, tzinfo=tzlocal()),
 'VALIDATED': datetime.datetime(2023, 1, 26, 15, 22, 6, 629000, tzinfo=tzlocal()),
 'QUEUED': datetime.datetime(2023, 1, 26, 15, 22, 6, 728000, tzinfo=tzlocal()),
 'RUNNING': datetime.datetime(2023, 1, 26, 15, 22, 7, 752000, tzinfo=tzlocal()),
 'COMPLETED': datetime.datetime(2023, 1, 26, 15, 22, 19, 271000, tzinfo=tzlocal())}
```
I think we can pick RUNNING time as a `run_time`, and separately add `created_time` to cover your case of reanalysis. Updated rfc with [9a62098](https://github.com/Qiskit/rfcs/pull/28/commits/9a62098d63278310863c9d1c34c431f4fa19cee7).